### PR TITLE
rpc: Implement `chainSpec` RPC API

### PR DESCRIPTION
**Pull request by lexnv**
This PR lays the ground for the implementation of the RPC spec
and adds the `chainSpec` methods.

The RPC methods are placed in the `rpc-spec` crate subject to
change in the future.

The `chainSpec` methods include:
 - `chainSpec_unstable_chainName` - Get the chain name, as present in the chain specification.
 - `chainSpec_unstable_genesisHash` - Get the chain's genesis hash.
 - `chainSpec_unstable_properties` - Get the properties of the chain, as present in the chain specification.

### Testing Done
- Unit tests are added for each methods
- CURL requests submitted to the local node

```bash
$ curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "chainSpec_unstable_properties"}' http://localhost:9933/
  {"jsonrpc":"2.0","result":{},"id":1}

curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "chainSpec_unstable_chainName"}' http://localhost:9933/
  {"jsonrpc":"2.0","result":"Development","id":1}

 curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "chainSpec_unstable_genesisHash"}' http://localhost:9933/
{"jsonrpc":"2.0","result":"0x7d341d884f3bea9b5668d29be58ac5e307d57122a7df21021a2099ffdd609928","id":1}
``` 

First part of #12071 .

### CODE REVIEW

File: client/rpc-spec/src/chain_spec/chain_spec.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+/// An API for chain spec RPC calls.
+pub struct ChainSpec {
+	/// The name of the chain.
+	name: String,
+	/// The hexadecimal encoded hash of the genesis block.
+	///
+	/// # Note
+	///
+	/// This can be empty, or prefixed with "0x".
```
Comment (tomaka) [inline client/rpc-spec/src/chain_spec/chain_spec.rs:33]: Why can this be empty? There's never a situation where we don't know the genesis hash.

Comment (lexnv) [reply client/rpc-spec/src/chain_spec/chain_spec.rs:33]: It makes sense, indeed I was thinking of this to receive untrusted/malformed input. Given we control that, will make sure to always format it as `0x[hash]`. Thanks for spotting this out!

Comment (lexnv) [reply client/rpc-spec/src/chain_spec/chain_spec.rs:33]: Have pushed a fix for it! Thanks again!

**Comment by tomaka** [comment]
The logic looks good to me. Can't comment on the code organization.

**Comment by niklasad1** [comment]
I think it's better to call the crate `rpc-spec` something like `rpc-api-v2/rpcv2` instead if we are planning to merge piece by piece.

For now it's probably alright to have the `rpc api def` and the implementation in the same crate as you did otherwise looks good.

File: client/rpc-spec/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! JSON-RPC interface exposed by blockchain nodes in the Substrate ecosystem.
+//!
+//! Specification [document](https://paritytech.github.io/json-rpc-interface-spec/).
```
Comment (niklasad1) [inline client/rpc-spec/src/lib.rs:21]: :+1:

File: client/rpc-spec/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! JSON-RPC interface exposed by blockchain nodes in the Substrate ecosystem.
```
Comment (niklasad1) [inline client/rpc-spec/src/lib.rs:19]:
```suggestion
//! Substrate JSON-RPC interface v2.
```

**Review by niklasad1 — APPROVED** [review]

**Review by skunert — APPROVED** [review]
Changes look good to me.

**Comment by lexnv** [comment]
We have also decided to remove the `version: 1` from the result of the `rpc_methods`. 
This is to simplify the `rpc_methods` response, as the new API should group and version methods.

File: client/rpc-spec-v2/src/chain_spec/api.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2022 Parity Technologies (UK) Ltd.
```
Comment (kianenigma) [inline client/rpc-spec-v2/src/chain_spec/api.rs:3]:
```suggestion
// Copyright (C) 2022 Parity Technologies (UK) Ltd.
```

File: client/rpc-spec-v2/src/chain_spec/api.rs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+use jsonrpsee::{core::RpcResult, proc_macros::rpc};
+use sc_chain_spec::Properties;
+
+#[rpc(client, server)]
+pub trait ChainSpecApi {
+	/// Get the chain name, as present in the chain specification.
+	///
+	/// # Unstable
+	///
+	/// This method is unstable and subject to change in the future.
```
Comment (kianenigma) [inline client/rpc-spec-v2/src/chain_spec/api.rs:30]: why?

File: client/rpc-spec-v2/src/chain_spec/chain_spec.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020-2022 Parity Technologies (UK) Ltd.
```
Comment (kianenigma) [inline client/rpc-spec-v2/src/chain_spec/chain_spec.rs:3]:
```suggestion
// Copyright (C) 2022 Parity Technologies (UK) Ltd.
```

**Review by kianenigma — APPROVED** [review]
LGTM except all the wrong license years.

File: client/rpc-spec-v2/src/chain_spec/api.rs (hunk 4 above)
Comment (lexnv) [reply client/rpc-spec-v2/src/chain_spec/api.rs]: Nice catch! Thanks!

File: client/rpc-spec-v2/src/chain_spec/api.rs (hunk 5 above)
Comment (lexnv) [reply client/rpc-spec-v2/src/chain_spec/api.rs:30]: Although they will be part of the final RPC, these methods are expected to go through a testing phase before standardizing and replacing the old API. During this time (1-2months TBD), we can still add breaking changes. At least, that's how I interpret the `unstable` from the [spec](https://paritytech.github.io/json-rpc-interface-spec/grouping-functions-and-node-capabilities.html)

File: client/rpc-spec-v2/src/chain_spec/api.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	#[method(name = "chainSpec_unstable_chainName")]
+	fn chain_spec_unstable_chain_name(&self) -> RpcResult<String>;
+
+	/// Get the chain's genesis hash.
+	///
+	/// # Unstable
+	///
+	/// This method is unstable and subject to change in the future.
+	#[method(name = "chainSpec_unstable_genesisHash")]
+	fn chain_spec_unstable_genesis_hash(&self) -> RpcResult<String>;
```
Comment (jsdw) [inline client/rpc-spec-v2/src/chain_spec/api.rs:40]: Tiny thing; I think of `chainspec` as one word personally; `chainspec_unstable_genesis_hash` for instance.

Not bothered either way though :)

**Review by jsdw — APPROVED** [review]

**Pull request merged**